addon/-private/properties/collection/main.js

```diff
@@ -0,0 +1,104 @@
+/* global Symbol */
+import Ember from 'ember';
+import { buildSelector, assign } from '../../helpers';
+import { create } from '../../create';
+import { count } from '../count';
+import Ceibo from 'ceibo';
+
+export class Collection {
+  constructor(definition, parent, key) {
+    this.definition = definition;
+    this.parent = parent;
+    this.key = key;
+
+    this._itemCounter = create({
+      count: count(definition.scope, {
+        resetScope: definition.resetScope,
+        testContainer: definition.testContainer
+      })
+    }, { parent });
+
+    this._items = [];
+  }
+
+  get length() {
+    return this._itemCounter.count;
+  }
+
+  objectAt(index) {
+    if (typeof this._items[index] === 'undefined') {
```

Comment:
What do you think of having this return `undefined` if `index >= this.length`?

Reply:
I like that idea, brings us inline with actual array semantics. Will update and add a test

Reply:
Actually, this does interfere with the current behavior which is based on the legacy collection's behavior, see [this test](https://github.com/pzuraq/ember-cli-page-object/blob/5f2bb1d67802faf94809d7b23071ef4ded535830/tests/unit/-private/properties/collection-test.js#L189-L203).

I'm open to either behavior being correct, we could also throw an error when they first try to get the page itself.

Reply:
Ah, yeah, I guess throwing an error makes the most sense, since that's what other page object methods do if you try to interact with an element that doesn't exist.

Reply:
Returning to this, I think an error may actually be too harsh of an assertion. I think your first instinct was right, returning `undefined` brings us inline with standard `objectAt` behavior

Reply:
Specifically, asseting prevents users from checking whether an item exists or not, and makes that case always an error, which I think could be really counterintuitive. Like if you just deleted an item and wanted to check if it was gone

Reply:
On the other hand, if you wanted to check if it was gone, you could check `length`. I guess if we want to be consistent with the expectations that `objectAt` sets up for developers, we should return `undefined`. If we want to be consistent with other page object helpers, we should throw an error. Does that seem like an accurate description of the decision we're facing? @san650, do you have a preference for what happens when you try to access an item in a collection that does not exist? Do you have a preference for whether we get `undefined` or an element-not-found error?

Reply:
@magistrula @pzuraq I think it should throw an element-not-found error. When you're writing tests, your asserting on things you know they should be true, so for me its weird to receive an `undefined`.

Reply:
I'm probably a little bit late to the party. Anyways..



When we try to access a regular page object property which doesn't exist in the DOM we don't throw any errors. Instead we return a page object node which allows `isPresent` and `isVisible` to be executed.

Reply:
> Specifically, asseting prevents users from checking whether an item exists or not, and makes that case always an error, which I think could be really counterintuitive

Agree

Reply:
In this case, it feels less to me like we're accessing a page object property that doesn't exist and more like we're accessing an element that doesn't exist. Thoughts?

Reply:
I think it's a bit of a paradigm shift, because before page objects always existed. The items that they referred to may-or-may not be in the DOM, but that didn't affect whether or not the page for the item existed - which is why you would have to assert `isPresent` on the individual item rather than checking if the page exists.

To be fair, I think that could trip up new users as well. Currently if we do `items(123456789)` it will return a page, has that ever thrown a user off?

Reply:
> Currently if we do `items(123456789)` it will return a page, has that ever thrown a user off?

I think there no any usages in the wild of `items(123456789)` exactly in this form. The user always wants to have an access to some page object property from the item:
```
  items(123456789).isVisible
  items(123456789).fillIn('something') // I think here we can assert as we do now
  items(123456789).title.click() // that's why `undefined` is not the best option
```

I think we already have a well defined behavior for all this cases and I find it pretty intuitive.